jdr/WFLY-13539-galleon_layer_jdr_support.adoc

```diff
@@ -0,0 +1,89 @@
+----
+[standalone@localhost:9990 /] /subsystem=jdr:generate-jdr-report()
+----
+
+Users are also able to generate the diagnostic information via command line executing the `jdr.sh` script located at `$WFLY_HOME/bin/` directory. This script accepts the following options:
+
+[source,bash]
+----
+$WFLY_HOME/bin/jdr.sh --help
+usage: jdr.{sh,bat,ps1} [options]
+
+JBoss Diagnostic Reporter (JDR) is a subsystem built to collect information to aid in troubleshooting. The jdr script is a utility for generating JDR reports.
+-c,--config <arg>     Configuration file of the server if it is not running.
+-h,--help             Display this message and exit
+-H,--host <arg>       hostname that the management api is bound to. (default: localhost)
+-p,--port <arg>       port that the management api is bound to. (default: 9990)
+-s,--protocol <arg>   Protocol that is used to connect. Can be remote, http or https (default: http)
+----
+
+The availability of a management interface is optional. If there is not a management interface available, the `jdr.sh` script is still able to extract some information invoking internally the embedded server.
+
+The goal of this feature is to supply a https://docs.wildfly.org/galleon/#_layers[Galleon layer] to give support to the JDR subsystem.
+
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.jboss.org/browse/WFLY-13539[WFLY-13539 Add a Galleon layer for JBoss Diagnostic Reporting]
+* https://issues.redhat.com/browse/EAP7-1444[EAP7-1444 Galleon Layers in EAP XP 2]
+
+=== Related Issues
+* https://issues.redhat.com/browse/WFGP-177[WFGP-177 Create a Galleon package for the JBoss Diagnostic Reporting bin scripts]
+
+=== Dev Contacts
+
+* mailto:brian.stansberry@redhat.com[Brian Stansberry]
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:fburzigo@redhat.com[Fabio Burzigotti]
+* mailto:jkasik@redhat.com[Jan Kasik]
+
+=== Testing By
+
+* [X] Engineering
+
+* [ ] QE
+
+=== Affected Projects or Components
+
+WildFly, WildFly Galleon Plugins
+
+=== Other Interested Projects
+
+== Requirements
+
+* The galleon layer must provide the provisioning of the JDR subsystem.
+* The galleon layer will depend optionally on `secure-management` layer. This layer will allow the users to configure a secure management interface to connect to the server via `jdr.sh`. If a user wants to connect via CLI, they have to add `core-tools` layer to their provision configuration.
+* The bash scripts to generate the diagnosis information must be provisioned and be available at `$WFLY_HOME/bin/`.
+
+== Test Plan
+
+The test coverage of the Galleon layer added by this proposal is divided in two main groups:
+
+. Test the Galleon layer provision. https://github.com/wildfly/wildfly/blob/master/testsuite/layers/src/test/java/org/jboss/as/test/layers/LayersTestCase.java[LayersTestCase] will be expanded to test the provision of the added layer when it is used isolated and with other layers. For each sort of provision, this test does the following:
```

Comment:
> I am wondering whether the wording should be Testing the Galleon layer provisioning. WDYT?

Replaced.

Yes, this description it is a general one. Specific details will be added on the test plan. The intention is to give an idea what kind of tests will be used. The smoke test executed by LayersTestCase and the execution of available tests ion WildFly test suite. My intention was not go into details of testing in this analysis doc. 

Do you agree? we will have the details in the test plan itself.

I've rephrased a bit these paragraphs, I hope it is more clear now. But let me know. They will be used in the other analysis docs.

In relation about how to link the wildfly/galleon-plugins#159 with a test, 
This [JdrReportManagmentTestCase](https://github.com/wildfly/wildfly/blob/master/testsuite/integration/basic/src/test/java/org/jboss/as/test/integration/jdr/mgmt/JdrReportManagmentTestCase.java) test will be expanded to verify:

. The presence of the shell scripts that allows a user to generate the JDR report via the command line.
. The validation of the content including more entries that are expected to be in the generated report.

This information is currently on the test plan itself.

Reply:
> Replaced.
Yes, this description it is a general one. Specific details will be added on the test plan. The intention is to give an idea what kind of tests will be used. The smoke test executed by LayersTestCase and the execution of available tests ion WildFly test suite. My intention was not go into details of testing in this analysis doc.
Do you agree? we will have the details in the test plan itself.

Sure I do agree. I was just wondering why https://github.com/wildfly/wildfly/blob/master/testsuite/layers/src/test/java/org/jboss/as/test/layers/LayersTestCase.java is not yet in the PR modified files. Am I missing something?

> I've rephrased a bit these paragraphs, I hope it is more clear now. But let me know. They will be used in the other analysis docs.

Sure, thanks. I'll have a look in a few minutes.

> In relation about how to link the wildfly/galleon-plugins#159 with a test,
This JdrReportManagmentTestCase test will be expanded to verify:
. The presence of the shell scripts that allows a user to generate the JDR report via the command line.
. The validation of the content including more entries that are expected to be in the generated report.
This information is currently on the test plan itself.

Yeah basically I was not understanding why `JdrReportManagmentTestCase` was in the PR while `LayersTestCase` wasn't but was mentioned here. Thanks for pointing me at the test plan.

Reply:
> I've rephrased a bit these paragraphs, I hope it is more clear now. But let me know. They will be used in the other analysis docs.

> Sure, thanks. I'll have a look in a few minutes.

BTW the wording seems appropriate to me, now, thanks!

Reply:
> BTW maybe I would:
> 
> * s/`with this layer isolated`/`with this layer in isolation`/.
> * s/`which are directly testing features added by this layer`/`which are directly testing this layer functionalities`/
>   WDYT?

Much better, thanks, I corrected them following the suggestion.

> And besides:, when you state:
> `this layer isolated and with this layer combined with all the layers`
> Do you mean with _all_ of the available layers? - i.e. this one + N-1 layers? Or pair combinations?

Yes, this must be fixed. There are some incompatible layers jpa + jpa-distributed as an example. The original list we have is [here](https://github.com/wildfly/wildfly/blob/master/testsuite/layers/pom.xml#L2067)

I'm going to try to get merged a Jira task with an additional list with more layers, and even better, having a section for jpa and jpa-distributed. Each new PR related to galleon layers will be based on the pre-existent list.

Thoughts?

Reply:
In relation to the above, I've just filed this Jira: https://issues.redhat.com/browse/WFLY-13703 That pretends to add all existing layers into a provision tested by LayersTestCase

Reply:
Hi @yersan - sorry for the late reply. I haven't checked the commits in WFLY-13703 in detail (BTW, thanks for the effort there) but it seems to me that now we'd have `LayersTestCase` now running with different configurations made of 1 (the currently tested) + 1 (chosen from the N-1 avaiable remaining) layers, right? So basically it will be testing 1+1 combinations?

Reply:
Hi @fabiobrz, with WFLY-13703  merged we are running `LayersTestCase` with the following three combinations:

* The current layer under test, hence `jdr`, provisioned in isolation.
* The current layer under test, hence `jdr`, provisioned with all the already available layers excluding `jpa-distribution` but including `jpa`.
* The current layer under test, hence `jdr`, provisioned with all the already available layers excluding `jpa` but including `jpa-distribution`

The total current list of available layers is [this](https://github.com/wildfly/wildfly/blob/master/testsuite/layers/pom.xml#L2147-L2199)

`jpa` and `jpa-distribution` are mutually exclusive, hence a separation in these two groups. We should do the same with the `ejb3-local-cache` and `ejb-dist-cache` once they are available.

Reply:
Ok, perfect, thanks for clarifying it. So I'd say it is confirmed that the layer combinations are always pairs - i.e. 1+1.

Reply:
Uhm, actually reading again:

`provisioned with all the already available layers excluding jpa-distribution but including jpa`

maybe this doesn't mean `1+1` but instead `1+N-1`, right?

Reply:
> maybe this doesn't mean 1+1 but instead 1+(N-1), right?

Yes `1+(N-1)`, being N all the layers, the list of all is [here](https://github.com/wildfly/wildfly/blob/master/testsuite/layers/pom.xml#L2147-L2199) 

Currently these layers cannot be provisioned at the same time `jpa-distribution` and `jpa`